### PR TITLE
Fix to handle Redfish Gen2 Firmware upgrade

### PR DESCRIPTION
**Pull request by cmadarsh**
##### SUMMARY
Fix to enable firmware upgrade on our Gen2 redfish based UltraStar Data102 platform.

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
wdc_redfish_command module

##### ADDITIONAL INFORMATION
Firmware upgrade in case of Gen2 UltraStar Data102 platform is handled with the new firmware bundle format.

### CODE REVIEW

**Comment by felixfontein** [comment]
Thanks for your contribution! Please check the failing sanity tests, and add a [changelog fragment](https://docs.ansible.com/ansible/devel/community/collection_development_process.html#creating-a-changelog-fragment). Thanks.

**Comment by cmadarsh** [comment]
Ready for review

File: changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - wdc_redfish_command - minor change to handle upgrade file for redfish gen 2 platform (https://github.com/ansible-collections/community.general/pull/8444).
```
Comment (felixfontein) [inline changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml:2]: I think Redfish is usually capitalized? (No idea about `gen 2`.)
```suggestion
  - wdc_redfish_command - minor change to handle upgrade file for Redfish gen 2 platform (https://github.com/ansible-collections/community.general/pull/8444).
```

**Review by felixfontein — COMMENTED** [review]
Should the documentation of the wdc_redfish_command module also be updated to mention this somehow?

File: changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml (hunk 1 above)
Comment (cmadarsh) [reply changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml:2]: Gen 2 upgraded version of WD Redfish based enclosure

**Comment by cmadarsh** [comment]
> Should the documentation of the wdc_redfish_command module also be updated to mention this somehow?

By default wdc_redfish_command module should work irrespective of the Redfish platform versions. I feel it need not be documented.

File: changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml (hunk 1 above)
Comment (cmadarsh) [reply changelogs/fragments/8444-fix-redfish-gen2-upgrade.yaml]: > I think Redfish is usually capitalized? (No idea about `gen 2`.)

Changes done and pushed.

**Comment by cmadarsh** [comment]
Ready for review

File: plugins/module_utils/wdc_redfish_utils.py [hunk 2]
```diff
@@ -180,33 +187,60 @@ def _get_bundle_version(self,
...
+                match_bin = re.match(pattern_bin, filename)
+                if match_bin is not None:
+                    bin_filename = match_bin.group(1)
+                    bin_file = tf.extractfile(bin_filename)
+                    bin_file.seek(11)
+                    byte_11 = bin_file.read(1)
+                    is_multi_tenant = byte_11 == b'\x80'
+                    gen = "G1"
+
+            return bundle_version, is_multi_tenant, gen
```
Comment (felixfontein) [inline plugins/module_utils/wdc_redfish_utils.py:243]: This way it's easier to read (less indentation) and the diff should be simpler:
```suggestion

        # Bundle is for MM or DP G1
        tf = tarfile.open(bundle_temp_filename)
        pattern_pkg = r"oobm-(.+)\.pkg"
        pattern_bin = r"(.*\.bin)"
        for filename in tf.getnames():
            match_pkg = re.match(pattern_pkg, filename)
            if match_pkg is not None:
                bundle_version = match_pkg.group(1)
            match_bin = re.match(pattern_bin, filename)
            if match_bin is not None:
                bin_filename = match_bin.group(1)
                bin_file = tf.extractfile(bin_filename)
                bin_file.seek(11)
                byte_11 = bin_file.read(1)
                is_multi_tenant = byte_11 == b'\x80'
                gen = "G1"
         return bundle_version, is_multi_tenant, gen
```

File: plugins/module_utils/wdc_redfish_utils.py [hunk 3]
```diff
@@ -180,33 +187,60 @@ def _get_bundle_version(self,
...
-
-        return bundle_version, is_multi_tenant
+        gen = None
+
+        # If not tarfile, then if the file has "MMG2" or "DPG2" at 2048th byte
+        # then the bundle is for MM or DP G2
+        if not tarfile.is_tarfile(bundle_temp_filename):
+            bundle_file = open(bundle_temp_filename, "rb")
+            bundle_file.seek(2048)
+            cookie1 = bundle_file.read(4)
```
Comment (felixfontein) [inline plugins/module_utils/wdc_redfish_utils.py:209]:
```suggestion
            with open(bundle_temp_filename, "rb") as bundle_file:
                bundle_file.seek(2048)
                cookie1 = bundle_file.read(4)
```
Also you should add a check whether the file is at least 2048 + 4 bytes long before trying to read at that position.

Comment (cmadarsh) [reply plugins/module_utils/wdc_redfish_utils.py]: Implemented the comments, please review

**Comment by felixfontein** [comment]
I cannot judge the actual change. If nobody objects, I'll merge this in 1-2 weeks.

**Comment by cmadarsh** [comment]
> I cannot judge the actual change. If nobody objects, I'll merge this in 1-2 weeks.

Ok, please merge, the changes are specific to Western Digital platform and its modules, which has been tested.

**Comment by felixfontein** [comment]
@cmadarsh thanks for your contribution!

**Pull request merged**